sdk/python/lib/pulumi/x/automation/local_workspace.py

```diff
@@ -0,0 +1,399 @@
+        path = os.path.join(self.work_dir, f"Pulumi{found_ext}")
+        with open(path, "w") as file:
+            json.dump(settings, file, indent=4) if found_ext == ".json" else yaml.dump(settings, stream=file)
+
+    def stack_settings(self, stack_name: str) -> StackSettings:
+        stack_settings_name = get_stack_settings_name(stack_name)
+        for ext in setting_extensions:
+            path = os.path.join(self.work_dir, f"Pulumi.{stack_settings_name}{ext}")
+            if not os.path.exists(path):
+                continue
+            with open(path, "r") as file:
+                settings = json.load(file) if ext == ".json" else yaml.safe_load(file)
+                return StackSettings(**settings)
+        raise FileNotFoundError(f"failed to find stack settings file in workdir: {self.work_dir}")
+
+    def save_stack_settings(self, stack_name: str, settings: StackSettings) -> None:
+        stack_settings_name = get_stack_settings_name(stack_name)
+        found_ext = ".yaml"
+        for ext in setting_extensions:
+            test_path = os.path.join(self.work_dir, f"Pulumi.{stack_settings_name}{ext}")
+            if os.path.exists(test_path):
+                found_ext = ext
+                break
+        path = os.path.join(self.work_dir, f"Pulumi.{stack_settings_name}{found_ext}")
+        with open(path, "w") as file:
+            json.dump(settings, file, indent=4) if found_ext == ".json" else yaml.dump(settings, stream=file)
+
+    def serialize_args_for_op(self, stack_name: str) -> List[str]:
+        # Not used by LocalWorkspace
+        return []
+
+    def post_command_callback(self, stack_name: str) -> None:
+        # Not used by LocalWorkspace
+        return
+
+    def get_config(self, stack_name: str, key: str) -> ConfigValue:
+        self.select_stack(stack_name)
+        result = self._run_pulumi_cmd_sync(["config", "get", key, "--json"])
+        val = json.loads(result.stdout)
+        return ConfigValue(**val)
+
+    def get_all_config(self, stack_name: str) -> ConfigMap:
+        self.select_stack(stack_name)
+        result = self._run_pulumi_cmd_sync(["config", "--show-secrets", "--json"])
+        config_json = json.loads(result.stdout)
+        config_map: ConfigMap = {}
+        for key in config_json:
+            config_map[key] = ConfigValue(**config_json[key])
+        return config_map
+
+    def set_config(self, stack_name: str, key: str, value: ConfigValue) -> None:
+        self.select_stack(stack_name)
+        secret_arg = "--secret" if value.secret else "--plaintext"
+        self._run_pulumi_cmd_sync(["config", "set", key, value.value, secret_arg])
+
+    def set_all_config(self, stack_name: str, config: ConfigMap) -> None:
+        config_adds = []
+        with futures.ThreadPoolExecutor() as executor:
+            for key in config:
+                config_adds.append(executor.submit(self.set_config, stack_name, key, config[key]))
+        futures.wait(config_adds)
```

Comment:
Why is the use of `ThreadPoolExecutor` and `futures.wait` necessary? `set_config` is currently synchronous, so couldn't it just be called in a loop?

```python
def set_all_config(self, stack_name: str, config: ConfigMap) -> None:
    for key in config:
        self.set_config(stack_name, key, config[key])
```

Reply:
`set_config` is synchronous but `set_all` should execute in bulk. Using the `ThreadPoolExecutor` lets each set_config run in parallel without making the user deal with the async/await interface.

It improved the test for [bulk_config_ops](https://github.com/pulumi/pulumi/pull/5979/files/dd8df192616f029cb2fdead2da291758d0efc848#diff-9fab3179db2be36998c2b9bd1a1fb1744888a8e4b99316d5872d5dafb5a00642R209-R231) from 50s to 12s. This has been a point of contention in our [nodejs implementation](https://github.com/pulumi/pulumi/issues/6042) as well.

Reply:
Never mind, going to do this in a loop until we sort out https://github.com/pulumi/pulumi/issues/6050